conversion/resource_generation.go

```diff
@@ -0,0 +1,413 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package conversion handles initial setup for the command line tool
+// and web APIs.
+
+// TODO:(searce) Organize code in go style format to make this file more readable.
+//
+//	public constants first
+//	key public type definitions next (although often it makes sense to put them next to public functions that use them)
+//	then public functions (and relevant type definitions)
+//	and helper functions and other non-public definitions last (generally in order of importance)
+package conversion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	datastream "cloud.google.com/go/datastream/apiv1"
+	"cloud.google.com/go/datastream/apiv1/datastreampb"
+	instance "cloud.google.com/go/spanner/admin/instance/apiv1"
+	"cloud.google.com/go/spanner/admin/instance/apiv1/instancepb"
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/internal"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/profiles"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/sources/common"
+	"github.com/google/uuid"
+	"google.golang.org/api/iterator"
+)
+
+type resourceGenerationInterface interface {
+	multiError(errorMessages []error) error
+	getConnProfilesRegion(ctx context.Context, projectId string, region string, dsClient *datastream.Client) ([]string, error)
+	connectionProfileExists(ctx context.Context, projectId string, profileName string, profileLocation string, connectionProfiles map[string][]string, dsClient *datastream.Client, s resourceGenerationInterface) (bool, error)
+	getResourcesForCreation(ctx context.Context, projectId string, sourceProfile profiles.SourceProfile, region string, validateOnly bool, dsClient *datastream.Client) ([]*ConnectionProfileReq, []*ConnectionProfileReq, error)
+	prepareMinimalDowntimeResources(createResourceData *ConnectionProfileReq, mutex *sync.Mutex) common.TaskResult[*ConnectionProfileReq]
+	setConnectionProfileFromRequest(details *ConnectionProfileReq, req *datastreampb.CreateConnectionProfileRequest) error
+	GetSpannerRegion(ctx context.Context, projectId string, instanceName string) (string, error)
+	connectionProfileCleanUp(ctx context.Context, profiles []*ConnectionProfileReq) error
+	ValidateResourceGeneration(ctx context.Context, projectId string, instanceId string, sourceProfile profiles.SourceProfile, conv *internal.Conv) error
+}
```

Comment:
Should the interface contain private methods?
Even in unit testing we do not typically mock private methods.

Reply:
Breaking this up in the new PR